From 885524c10f441321003c7f809b0d4e8747fda3a9 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Mon, 14 Apr 2014 09:46:20 +1000
Subject: [PATCH] Use Provided Case for Non Existent Offline Players


diff --git a/src/main/java/net/minecraft/server/players/UserCache.java b/src/main/java/net/minecraft/server/players/UserCache.java
index b9ecd819b6..5ca7755d5c 100644
--- a/src/main/java/net/minecraft/server/players/UserCache.java
+++ b/src/main/java/net/minecraft/server/players/UserCache.java
@@ -140,7 +140,7 @@ public class UserCache {
             usercache_usercacheentry.setLastAccess(this.getNextOperation());
             optional = Optional.of(usercache_usercacheentry.getProfile());
         } else {
-            optional = lookupGameProfile(this.profileRepository, s1);
+            optional = lookupGameProfile(this.profileRepository, s); // Spigot - use correct case for offline players
             if (optional.isPresent()) {
                 this.add((GameProfile) optional.get());
                 flag = false;
-- 
2.42.0

